File: composeApp/src/androidMain/kotlin/com/zrifapps/goservice/ui/vehicle/components/VehicleForm.kt
package com.zrifapps.goservice.ui.vehicle.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.zrifapps.goservice.ui.components.AppTextField
import com.zrifapps.goservice.ui.theme.AppColors
import com.zrifapps.goservice.ui.theme.FaIcon
import com.zrifapps.goservice.ui.theme.FaIcons
import com.zrifapps.goservice.ui.theme.FaStyle
import com.zrifapps.goservice.ui.theme.plusJakartaSansFontFamily

private val vehicleColors = listOf("#1A2418", "#D6453A", "#3FB1D6", "#E89C2E", "#F2EEE6", "#7B6FE8")

private fun hexToColor(hex: String): Color = try {
    Color(android.graphics.Color.parseColor(hex))
} catch (e: Exception) {
    Color.Gray
}

data class VehicleFormState(
    val type: String = "motor",
    val nama: String = "",
    val merek: String = "",
    val model: String = "",
    val tahun: String = "",
    val platNomor: String = "",
    val odometer: String = "",
    val warna: String = "#1A2418",
) {
    val isValid: Boolean get() = nama.isNotBlank() && merek.isNotBlank()
}

@Composable
fun VehicleForm(
    state: VehicleFormState,
    onStateChange: (VehicleFormState) -> Unit,
    modifier: Modifier = Modifier,
    showTypeSelector: Boolean = true,
) {
    val font = plusJakartaSansFontFamily()

    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(16.dp),
    ) {
        if (showTypeSelector) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(10.dp),
            ) {
                listOf("motor" to FaIcons.MOTORCYCLE, "mobil" to FaIcons.CAR).forEach { (tp, icon) ->
                    val selected = state.type == tp
                    Box(
                        modifier = Modifier
                            .weight(1f)
                            .clip(RoundedCornerShape(18.dp))
                            .background(if (selected) AppColors.Primary else AppColors.Surface)
                            .then(
                                if (!selected) Modifier.border(1.5.dp, AppColors.Border, RoundedCornerShape(18.dp))
                                else Modifier
                            )
                            .clickable { onStateChange(state.copy(type = tp)) }
                            .padding(16.dp),
                        contentAlignment = Alignment.Center,
                    ) {
                        Column(
                            horizontalAlignment = Alignment.CenterHorizontally,
                            verticalArrangement = Arrangement.spacedBy(6.dp),
                        ) {
                            FaIcon(
                                icon = icon,
                                color = if (selected) Color.White else AppColors.TextPrimary,
                                size = 28.sp,
                                style = FaStyle.Solid,
                            )
                            Text(
                                text = if (tp == "motor") "Motor" else "Mobil",
                                fontSize = 13.sp,
                                fontWeight = FontWeight.Bold,
                                fontFamily = font,
                                color = if (selected) Color.White else AppColors.TextPrimary,
                            )
                        }
                    }
                }
            }
        }

        AppTextField(
            label = "Nama panggilan",
            value = state.nama,
            onValueChange = { onStateChange(state.copy(nama = it)) },
            placeholder = "cth. Beat Hitam",
        )

        AppTextField(
            label = "Merk",
            value = state.merek,
            onValueChange = { onStateChange(state.copy(merek = it)) },
            placeholder = "cth. Honda",
        )

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            AppTextField(
                label = "Model",
                value = state.model,
                onValueChange = { onStateChange(state.copy(model = it)) },
                placeholder = "cth. Beat",
                modifier = Modifier.weight(1f),
            )
            AppTextField(
                label = "Tahun",
                value = state.tahun,
                onValueChange = { onStateChange(state.copy(tahun = it)) },
                placeholder = "cth. 2022",
                keyboardType = KeyboardType.Number,
                modifier = Modifier.weight(1f),
            )
        }

        AppTextField(
            label = "Plat nomor",
            value = state.platNomor,
            onValueChange = { onStateChange(state.copy(platNomor = it)) },
            placeholder = "cth. B 1234 ABC",
        )

        AppTextField(
            label = "KM saat ini",
            value = state.odometer,
            onValueChange = { onStateChange(state.copy(odometer = it)) },
            placeholder = "cth. 18000",
            keyboardType = KeyboardType.Number,
            imeAction = ImeAction.Done,
        )

        Column {
            Text(
                text = "Warna",
                fontSize = 13.sp,
                fontWeight = FontWeight.SemiBold,
                fontFamily = font,
                color = AppColors.TextMuted,
                modifier = Modifier.padding(bottom = 8.dp),
            )
            Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                vehicleColors.forEach { hex ->
                    val selected = state.warna == hex
                    Box(
                        modifier = Modifier
                            .size(36.dp)
                            .clip(RoundedCornerShape(12.dp))
                            .background(hexToColor(hex))
                            .then(
                                if (selected) Modifier.border(3.dp, AppColors.Primary, RoundedCornerShape(12.dp))
                                else Modifier.border(1.dp, AppColors.Border, RoundedCornerShape(12.dp))
                            )
                            .clickable { onStateChange(state.copy(warna = hex)) },
                    )
                }
            }
        }
    }
}

File: iosApp/iosApp/Features/Vehicle/Components/VehicleForm.swift
import SwiftUI
import Shared

private let vehicleColors = ["#1A2418", "#D6453A", "#3FB1D6", "#E89C2E", "#F2EEE6", "#7B6FE8"]

private extension Color {
    init(hex: String) {
        let hex = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int: UInt64 = 0
        Scanner(string: hex).scanHexInt64(&int)
        let r = Double((int >> 16) & 0xFF) / 255
        let g = Double((int >> 8) & 0xFF) / 255
        let b = Double(int & 0xFF) / 255
        self.init(red: r, green: g, blue: b)
    }
}

struct VehicleFormState {
    var type: String = "motor"
    var subtype: String = VehicleSubtypes.defaultFor(vehicleType: "motor")
    var nama: String = ""
    var merek: String = ""
    var model: String = ""
    var tahun: String = ""
    var platNomor: String = ""
    var odometer: String = ""
    var warna: String = "#1A2418"

    var isValid: Bool {
        !nama.trimmingCharacters(in: .whitespaces).isEmpty &&
        !merek.trimmingCharacters(in: .whitespaces).isEmpty
    }
}

extension VehicleFormState {
    func toOnboardingInput(typeKey: String? = nil) -> OnboardingVehicleInput {
        let parsedYear: KotlinInt? = Int32(tahun).map { KotlinInt(value: $0) }
        return PresentationFactory.shared.vehicleInput(
            typeKey: typeKey ?? type,
            nickname: nama,
            brand: merek,
            model: model,
            year: parsedYear,
            plateNumber: platNomor,
            odometerKm: Int64(odometer) ?? 0,
            colorHex: warna,
            subtypeId: subtype.isEmpty ? nil : subtype
        )
    }
}

extension VehicleSubtypes {
    static func defaultFor(vehicleType: String) -> String {
        vehicleType == "mobil" ? "mpv" : "matic"
    }
}

struct VehicleForm: View {
    @Binding var state: VehicleFormState
    var showTypeSelector: Bool = true

    @State private var showSubtypePicker: Bool = false

    private var subtypeIcon: String {
        state.type == "mobil" ? "\u{f1b9}" : "\u{f21c}"
    }

    private var subtypeLabel: String {
        VehicleSubtypes.label(for: state.type, id: state.subtype)
    }

    private var subtypeOptions: [ActionSheetOption] {
        VehicleSubtypes.list(for: state.type).map { sub in
            ActionSheetOption(
                id: sub.id,
                iconUnicode: subtypeIcon,
                label: sub.label,
                subtitle: sub.desc,
                value: sub.id
            )
        }
    }

    var body: some View {
        VStack(spacing: 16) {
            if showTypeSelector {
                HStack(spacing: 10) {
                    ForEach([("motor", "\u{f21c}"), ("mobil", "\u{f1b9}")], id: \.0) { tp, icon in
                        let selected = state.type == tp
                        Button {
                            if state.type != tp {
                                state.type = tp
                                state.subtype = VehicleSubtypes.defaultFor(vehicleType: tp)
                            }
                        } label: {
                            VStack(spacing: 6) {
                                Text(icon)
                                    .font(.custom("FontAwesome6Free-Solid", size: 28))
                                    .foregroundColor(selected ? .white : .sgTextPrimary)
                                Text(tp == "motor" ? "Motor" : "Mobil")
                                    .font(.custom("PlusJakartaSans-Bold", size: 13))
                                    .foregroundColor(selected ? .white : .sgTextPrimary)
                            }
                            .frame(maxWidth: .infinity)
                            .padding(16)
                            .background(selected ? Color.sgPrimary : Color.white)
                            .clipShape(RoundedRectangle(cornerRadius: 18))
                            .overlay(
                                RoundedRectangle(cornerRadius: 18)
                                    .strokeBorder(selected ? Color.clear : Color.black.opacity(0.08), lineWidth: 1.5)
                            )
                        }
                        .buttonStyle(.plain)
                    }
                }
            }

            VStack(alignment: .leading, spacing: 6) {
                Text("Sub-tipe")
                    .font(.custom("PlusJakartaSans-SemiBold", size: 13))
                    .foregroundColor(.sgTextMuted)
                Button {
                    showSubtypePicker = true
                } label: {
                    HStack(spacing: 10) {
                        Text(subtypeIcon)
                            .font(.custom("FontAwesome6Free-Solid", size: 16))
                            .foregroundColor(.sgTextMuted)
                        Text(subtypeLabel)
                            .font(.custom("PlusJakartaSans-SemiBold", size: 15))
                            .foregroundColor(.sgTextPrimary)
                        Spacer()
                        Text("\u{f078}")
                            .font(.custom("FontAwesome6Free-Solid", size: 12))
                            .foregroundColor(.sgTextSubtle)
                    }
                    .padding(.horizontal, 16)
                    .padding(.vertical, 14)
                    .background(Color.white)
                    .clipShape(RoundedRectangle(cornerRadius: 14))
                    .overlay(
                        RoundedRectangle(cornerRadius: 14)
                            .strokeBorder(Color.black.opacity(0.08), lineWidth: 1.5)
                    )
                }
                .buttonStyle(.plain)
            }

            AppTextField(label: "Nama panggilan", placeholder: "cth. Beat Hitam", text: $state.nama)
            AppTextField(label: "Merk", placeholder: "cth. Honda", text: $state.merek)

            HStack(spacing: 12) {
                AppTextField(label: "Model", placeholder: "cth. Beat", text: $state.model)
                AppTextField(label: "Tahun", placeholder: "cth. 2022", text: $state.tahun, keyboardType: .numberPad)
            }

            AppTextField(label: "Plat nomor", placeholder: "cth. B 1234 ABC", text: $state.platNomor)
            AppTextField(label: "KM saat ini", placeholder: "cth. 18000", text: $state.odometer, keyboardType: .numberPad)

            VStack(alignment: .leading, spacing: 8) {
                Text("Warna")
                    .font(.custom("PlusJakartaSans-SemiBold", size: 13))
                    .foregroundColor(.sgTextMuted)
                HStack(spacing: 8) {
                    ForEach(vehicleColors, id: \.self) { hex in
                        let selected = state.warna == hex
                        RoundedRectangle(cornerRadius: 12)
                            .fill(Color(hex: hex))
                            .frame(width: 36, height: 36)
                            .overlay(
                                RoundedRectangle(cornerRadius: 12)
                                    .strokeBorder(
                                        selected ? Color.sgPrimary : Color.black.opacity(0.08),
                                        lineWidth: selected ? 3 : 1
                                    )
                            )
                            .onTapGesture { state.warna = hex }
                    }
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
        }
        .sheet(isPresented: $showSubtypePicker) {
            ActionSheetView(
                title: "Sub-tipe kendaraan",
                subtitle: "Pilih yang paling mendekati — menentukan komponen yang relevan.",
                options: subtypeOptions,
                selectionMode: .radio,
                selectedValue: state.subtype,
                onSelect: { value in
                    state.subtype = value
                }
            )
            .presentationDetents([.medium, .large])
            .presentationDragIndicator(.visible)
        }
    }
}
